### PR TITLE
feat(Guild): add Guild.fetchVanityData

### LINKED ISSUE #4061: Add function to get vanity URL uses
**Is your feature request related to a problem? Please describe.**
I want to get the uses of a vanity URL, but there is no function for that.

**Describe the ideal solution**
Add a function `Guild.fetchVanityUses()` that gets the uses of a vanity URL, if any.

**Describe alternatives you've considered**
It is already possible to get vanity invites via `client.fetchInvite(server.vanityURLCode)`, but the `.uses` is always `null`. Instead of the extra function the implementation could fill that field. Alternatively the `fetchInvites` function could be changed to include the vanity URL invite and its uses.

**Additional context**
The Discord API supports this: https://discordapp.com/developers/docs/resources/guild#get-guild-vanity-url
There is this code in https://github.com/discordjs/discord.js/blob/master/src/structures/Guild.js#L726-L746:

```js
  fetchVanityCode() {
    if (!this.features.includes('VANITY_URL')) {
      return Promise.reject(new Error('VANITY_URL'));
    }
    return this.client.api
      .guilds(this.id, 'vanity-url')
      .get()
      .then(res => res.code);
  }
```
so it should be relatively easy to add
```js
  fetchVanityUses() {
    if (!this.features.includes('VANITY_URL')) {
      return Promise.reject(new Error('VANITY_URL'));
    }
    return this.client.api
      .guilds(this.id, 'vanity-url')
      .get()
      .then(res => res.uses);
  }
```

### PR DESCRIPTION
**Pull request by CyberiumShadow**
**Please describe the changes this PR makes and why it should be merged:**

This PR adds a method `fetchVanityData` which returns the Vanity Invite Code and Uses from the API.

This also deprecates `fetchVanityCode`.

This should resolve #4061 

Talking with @vladfrangu on Discord, he advised me to implement `fetchVanityData` and deprecate `fetchVanityCode` instead of implementing `fetchVanityUses` as per #4061

**Status**

- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**

- [x] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by CyberiumShadow** [comment]
I think the only thing missing here is a `this.vanityURLUses` property but that require using the `fetchVanityData` method in this case to populate it as only `vanity_url_code` is included in the guild packet from `GUILD_CREATE`

**Comment by ItsPapaia** [comment]
`.then(res => res)` is useless ¯\_(ツ)_/¯

Also, that `{ code: ?string, uses: ?number }` is not quite right, make a typedef.

File: src/structures/Guild.js [hunk 1]
```diff
@@ -745,6 +746,28 @@ class Guild extends Base {
...
+   *   .catch(console.error);
+   */
+  fetchVanityData() {
+    if (!this.features.includes('VANITY_URL')) {
+      return Promise.reject(new Error('VANITY_URL'));
+    }
+    return this.client.api
+      .guilds(this.id, 'vanity-url')
+      .get()
+      .then(res => res);
```
Comment (kyranet) [inline src/structures/Guild.js:768]: 🧹
```suggestion
      .get();
```

**Comment by ItsPapaia** [comment]
Thanks @kyranet, didn't manage to do it myself 😄.

**Comment by CyberiumShadow** [comment]
> `.then(res => res)` does nothing.
> A typedef should be used instead of `{ code: ?string, uses: ?number }`.
> You deprecate it without giving a warning.
> You should set `vanityURLUses` when `fetchVanityData` is called.

In the case of deprecation, do you mean remove the `@deprecated` tag or add a description/reason to the tag.

And in the case of setting `vanityURLUses` should that be set explicitly when `fetchVanityData` is called or when the guild is setup from the `GUILD_CREATE`

**Comment by ItsPapaia** [comment]
> In the case of deprecation, do you mean remove the `@deprecated` tag or add a description/reason to the tag.

No, I mean you need to emit a deprecation warning (to the process)

> And in the case of setting `vanityURLUses` should that be set explicitly when `fetchVanityData` is called or when the guild is setup from the `GUILD_CREATE`

When setup it should be null (above the null assignment you'll place the JSDoc for it), and when calling `fetchVanityData`, you set the prop to the actual count.

**Comment by CyberiumShadow** [comment]
> > In the case of deprecation, do you mean remove the `@deprecated` tag or add a description/reason to the tag.
> 
> No, I mean you need to emit a deprecation warning (to the process)
> 
> > And in the case of setting `vanityURLUses` should that be set explicitly when `fetchVanityData` is called or when the guild is setup from the `GUILD_CREATE`
> 
> When setup it should be null (above the null assignment you'll place the JSDoc for it), and when calling `fetchVanityData`, you set the prop to the actual count.

Correct me if I am wrong, but I would be using `process.emitWarning` for this yes?

**Comment by ItsPapaia** [comment]
> Correct me if I am wrong, but I would be using process.emitWarning for this yes?

I am pretty sure the answer is yes, I'll ask the Discord.

**Comment by ItsPapaia** [comment]
Wait, @appellation says [util.deprecate](https://discordapp.com/channels/222078108977594368/222197033908436994/701478651039842334).

File: src/structures/Guild.js [hunk 2]
```diff
@@ -736,13 +743,44 @@ class Guild extends Base {
...
+      }
+      return this.client.api
+        .guilds(this.id, 'vanity-url')
+        .get()
+        .then(res => res.code);
+    }, 'fetchVanityCode() is deprecated. Use fetchVanityData() instead.');
+  }
+
+  /**
+   * An object containing information about a guild member's ban.
```
Comment (NotSugden) [inline src/structures/Guild.js:757]: this should be changed, its definitely not information about a guild member's ban

Comment (CyberiumShadow) [reply src/structures/Guild.js:757]: Lol oops, I overlooked that

File: src/structures/Guild.js [hunk 3]
```diff
@@ -745,6 +753,35 @@ class Guild extends Base {
...
+   * @returns {Promise<VanityData>}
+   * @example
+   * // Fetch invite data
+   * guild.fetchVanityData()
+   *   .then(res => {
+   *     console.log(`Vanity URL: https://discord.gg/${res.code}`);
+   *   })
+   *   .catch(console.error);
+   */
+  fetchVanityData() {
```
Comment (ItsPapaia) [inline src/structures/Guild.js:775]:
```suggestion
  async fetchVanityData() {
```

File: src/structures/Guild.js [hunk 4]
```diff
@@ -745,6 +753,35 @@ class Guild extends Base {
...
+    if (!this.features.includes('VANITY_URL')) {
+      return Promise.reject(new Error('VANITY_URL'));
+    }
+    const data = this.client.api.guilds(this.id, 'vanity-url').get();
```
Comment (ItsPapaia) [inline src/structures/Guild.js:779]:
```suggestion
    const data = await this.client.api.guilds(this.id, 'vanity-url').get();
```

File: src/structures/Guild.js [hunk 5]
```diff
@@ -311,6 +312,12 @@ class Guild extends Base {
      */
     this.vanityURLCode = data.vanity_url_code;
 
+    /**
+     * The use count of the vanity URL code of the guild, if any
```
Comment (NotSugden) [inline src/structures/Guild.js:316]: should there be a <warn> (or <info>) tag here to say it must have been fetched once?

Comment (CyberiumShadow) [reply src/structures/Guild.js:316]: For this, it's only populated on Guild#fetchVanityData but adding that to the @link pushes it over jsdoc eslint rules.

Should I just disable the rule for tha that line?

File: src/structures/Guild.js [hunk 6]
```diff
@@ -745,6 +753,35 @@ class Guild extends Base {
...
+   * // Fetch invite data
+   * guild.fetchVanityData()
+   *   .then(res => {
+   *     console.log(`Vanity URL: https://discord.gg/${res.code} with ${res.uses} uses`);
+   *   })
+   *   .catch(console.error);
+   */
+  async fetchVanityData() {
+    if (!this.features.includes('VANITY_URL')) {
+      return Promise.reject(new Error('VANITY_URL'));
```
Comment (vladfrangu) [inline src/structures/Guild.js:777]:
```suggestion
      throw new Error('VANITY_URL');
```

File: src/structures/Guild.js (hunk 5 above)
Comment (SpaceEEC) [reply src/structures/Guild.js:316]: Yes, disabling it is fine then.

**Comment by CyberiumShadow** [comment]
> the following line needs to be added at the top:
> 
> ```js
> const { Error } = require('../errors');
> ```
> 
> currently it throws a literal `VANITY_URL` error.
> 
> image

Is it specifically needed? 

All the other methods (leave and fetchBan) that throw an Error seem to be fine without

File: src/structures/Guild.js [hunk 7]
```diff
@@ -737,14 +749,43 @@ class Guild extends Base {
    */
   fetchVanityCode() {
     if (!this.features.includes('VANITY_URL')) {
-      return Promise.reject(new Error('VANITY_URL'));
+      throw new Error('VANITY_URL');
```
Comment (vladfrangu) [inline src/structures/Guild.js:752]: As this isn't an async function, if you use throw, you'll get a sync error not an async one
```suggestion
      return Promise.reject(new Error('VANITY_URL'));
```

**Comment by vladfrangu** [comment]
> > the following line needs to be added at the top:
> > ```js
> > const { Error } = require('../errors');
> > ```
> > 
> > 
> > currently it throws a literal `VANITY_URL` error.
> > image
> 
> Is it specifically needed?
> 
> All the other methods (leave and fetchBan) that throw an Error seem to be fine without

It is necessary for custom d.js errors that are in the [messages](https://github.com/discordjs/discord.js/blob/master/src/errors/Messages.js) file

File: src/structures/Guild.js [hunk 8]
```diff
@@ -1,12 +1,14 @@
...
 
+const { deprecate } = require('util');
 const Base = require('./Base');
 const GuildAuditLogs = require('./GuildAuditLogs');
 const GuildPreview = require('./GuildPreview');
 const Integration = require('./Integration');
 const Invite = require('./Invite');
 const VoiceRegion = require('./VoiceRegion');
 const Webhook = require('./Webhook');
+const { Error } = require('../errors');
```
Comment (vladfrangu) [inline src/structures/Guild.js:11]: I know nitpick, but can you re-add these back? Someone removed them at some point
```suggestion
const { Error, TypeError } = require('../errors');
```

File: src/structures/Guild.js [hunk 9]
```diff
@@ -1,12 +1,14 @@
...
 
+const { deprecate } = require('util');
 const Base = require('./Base');
 const GuildAuditLogs = require('./GuildAuditLogs');
 const GuildPreview = require('./GuildPreview');
 const Integration = require('./Integration');
 const Invite = require('./Invite');
 const VoiceRegion = require('./VoiceRegion');
 const Webhook = require('./Webhook');
+const { Error, TypeError } = require('../errors');
```
Comment (ItsPapaia) [inline src/structures/Guild.js:11]:
```suggestion
const { Error } = require('../errors');
```
-1 unused var

Comment (vladfrangu) [reply src/structures/Guild.js:11]: It is used in addMember

**Pull request merged**